src/include/sof/preproc.h

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2016, Intel Corporation
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Michal Jerzy Wierzbicki <michalx.wierzbicki@intel.com>
+ */
+
+#ifndef __INCLUDE_MACRO_METAPROGRAMMING__
+#define __INCLUDE_MACRO_METAPROGRAMMING__
+
+/* count number of var args - during preprocesing
+ * works for predefined number of args
+ * PP_NARG_BEFORE_COMPILE(A,B,C,D) evaluates to 4
+ */
+#define PP_NARG_BEFORE_COMPILE(...) IF_ELSE(HAS_ARGS(__VA_ARGS__)) (\
+		PP_NARG_BEFORE_COMPILE_(__VA_ARGS__, PP_RSEQ_N()) )\
+		(0)
+#define PP_NARG_BEFORE_COMPILE_(...) \
+		PP_ARG_N(__VA_ARGS__)
+#define PP_ARG_N(\
+	_1,   _2,  _3,  _4,  _5,  _6,  _7,  _8,  _9, _10, \
+	_11, _12, _13, _14, _15, _16, _17, _18, _19, _20, \
+	_21, _22, _23, _24, _25, _26, _27, _28, _29, _30, \
+	_31, _32, _33, _34, _35, _36, _37, _38, _39, _40, \
+	_41, _42, _43, _44, _45, _46, _47, _48, _49, _50, \
+	_51, _52, _53, _54, _55, _56, _57, _58, _59, _60, \
+	_61, _62, _63, N, ...) N
+#define PP_RSEQ_N() \
+	63, 62, 61, 60, \
+	59, 58, 57, 56, 55, 54, 53, 52, 51, 50, \
+	49, 48, 47, 46, 45, 44, 43, 42, 41, 40, \
+	39, 38, 37, 36, 35, 34, 33, 32, 31, 30, \
+	29, 28, 27, 26, 25, 24, 23, 22, 21, 20, \
+	19, 18, 17, 16, 15, 14, 13, 12, 11, 10, \
+	 9,  8,  7,  6,  5,  4,  3,  2,  1,  0
+
+/* treat x as string while forcing x expansion beforehand */
+#define _QUOTE(x) #x
+#define QUOTE(x) _QUOTE(x)
```

Comment:
@mwierzbix one more thing. It's probably better to split the macro into private and public. I guess you are already doing this with a _ prefix ?

Reply:
@lgirdwood you mean to split them into header (public macros) and source (private macros) files?
I don't know of any other way of separating access to macros.
Or you mean to `#undef` the "private" ones at the end of the file?

Not mentioning that from what I know (and read https://en.wikibooks.org/wiki/C_Programming/Language_Reference) `private` and `public` are C++/Java/C#/OOP concepts, not C.

Or do you mean https://en.wikibooks.org/wiki/C_Programming/Preprocessor_directives_and_macros#X-Macros ?

Reply:
No, I'm thinking we use a prefix like MP_ (for metaprogramming) or similar for all public macros and prefix of _ for all private macros, The private macros could either go at the top of the file or in another header with a -priv.h name suffix and a big warning about not to be included directly. This would keep things simple for users and maintainers.

Reply:
Seems fair enough.

Reply:
@lgirdwood division into "private" and "public" macros done.
Additionally added/reworked:
* unit tests to cover more advanced and less obvious macros,
and made DEFER macros more flexible
* #ifdef #error guards for including "private" macros outside of "public" macros usage

Reply:
I very much hope that these were the last changes here.
In case of any whitespace / indentation / formatting stuff, I'd rather fix them in separate PR.